src/lib/arch/riscv/ExceptionHandler.cc

```diff
@@ -633,14 +634,71 @@ bool ExceptionHandler::init() {
         stateChange = {ChangeType::REPLACEMENT, {R0}, {0ull}};
         break;
       }
-
       default:
         printException(instruction_);
         std::cout << "\n[SimEng:ExceptionHandler] Unrecognised syscall: "
                   << syscallId << std::endl;
         return fatal();
     }
 
+    return concludeSyscall(stateChange);
+  } else if (exception == InstructionException::AtomicOperation) {
+    // NOTE this may complete in only a single cycle and not the value from the
+    // config file. The instruction should be completely reissued and allowed to
+    // execute with latency once it reaches the head of the ROB
+
+    // Retrieve metadata, operand values and destination registers from
+    // instruction
+    auto metadata = instruction_.getMetadata();
+    auto operands = instruction_.getSourceOperands();
+    auto destinationRegs = instruction_.getDestinationRegisters();
+
+    ProcessStateChange stateChange;
+    switch (instruction_.getMetadata().opcode) {
+      case Opcode::RISCV_CSRRW:  // CSRRW rd,csr,rs1
+        if (metadata.operands[1].reg == RISCV_SYSREG_FRM) {
+          // Update CPP rounding mode but not floating point CSR as currently no
+          // implementation
+
+          switch (operands[0].get<uint64_t>()) {
+            case 0:  // RNE, Round to nearest, ties to even
+              fesetround(FE_TONEAREST);
+              break;
+            case 1:  // RTZ Round towards zero
+              fesetround(FE_TOWARDZERO);
+              break;
+            case 2:  // RDN Round down (-infinity)
+              fesetround(FE_DOWNWARD);
+              break;
+            case 3:  // RUP Round up (+infinity)
+              fesetround(FE_UPWARD);
+              break;
+            case 4:  // RMM Round to nearest, ties to max magnitude
+              // FE_TONEAREST ties towards even but no other options available
+              // in fenv
+              fesetround(FE_TONEAREST);
+              break;
+            default:
+              // Invalid Case
```

Comment:
The first part of the comment is a quote from the spec, I've added some quotes to try and make this more clear but can add more info if needed. The second part of the comment is a little more explanation and the solution. Nothing should happen when an incorrect dynamic rounding mode is set, an exception should only be raised if another instruction attempts to use it. The implementation I have in mind requires being able to set the CSR (which hasn't been implemented yet) so the comment serves as a reminder to implement this later.

In any case, it would require very poor code generation from the compiler to actually hit this case.

Reply:
That makes more sense now, thank you for the clarification.